### ISSUE TITLE
`no-descending-specificity` is very confusing when it comes to nested rules

### ISSUE THREAD

**Issue by clarfonthey** [stylelint/stylelint#7844]
### What is the problem you're trying to solve?

Here's a simple example of an error from one of my more complex stylesheets:

```
Expected selector "span::after" to come before selector "nav form label:first-of-type:hover span::after"
```

The relevant lines of the styles:

```
@media screen {
  // ...
  nav {
    // ...

    form {
      // ...

      label:first-of-type {
        // ...
        &:hover span::after { // complaining about rule here
          // ...
        }
      }

      label:last-of-type {
        span::after { // error here
          // ...
        }
      }
    }
  }
}
```

Now, even if it were to expand out the nested selector properly, like below:

```
Expected selector "nav form label:last-of-type span::after" to come before selector "nav form label:first-of-type:hover span::after"
```

It's clear that it's descending specificity, but not clear if:

1. This is even a problem when it comes to nested selectors.
2. How you'd ever fix this issue.

If you're interested, the full code is here: https://codeberg.org/clarfonthey/ltdk-xyz/src/branch/main/sass/_screen.scss#L100-L115

The point is, this code triggers often for nested selectors (which often only sort by specificity at a single nesting level) and it's unclear what to do about it.

To also describe what this code is doing: I have some forms in an options menu that are displayed as ASCII art entirely via CSS `::before` and `::after` content which can be ignored by screen readers and other assistive tech. The logic for verifying they show up correctly is a bit complicated, and the selectors are nested and organised by the order they show up on the page to make the code as readable as possible.

I'm including a more complex example to demonstrate something where simply expanding out the nested CSS would not be particularly helpful, rather than something simple which could do that instead to resolve the error.

### What solution would you like to see?

Personally, I'm not sure, but I wanted to point this out since `no-descending-specificity` shows up often for my nested rules. In general, I think that at least one of the following should apply:

* `no-descending-specificity` should only consider selectors at the same nesting level by default.
* `no-descending-specificity` should expand the full selectors in its errors, and point out the line number for conflicting selectors.
* `no-descending-specificity` should print a warning or error when used with nested rules.

And since I don't see any other issues pointing this out, I wanted to write one up to talk about it.

**Comment by Mouvedia** [comment]
> ### What solution would you like to see?
> …
> 1. `no-descending-specificity` should only consider selectors at the same nesting level by default.

This sounds like a request for new `depth` option.
i.e. _type: new option_

> 2. `no-descending-specificity` should expand the full selectors in its errors, and point out the line number for conflicting selectors.

That would be a _type: enhancement_.

> 3. `no-descending-specificity` should not be enabled by default.

Since I don't have access to your private registry I can't tell which config you are referring to.
The issue would be transferred to `stylelint-config-recommended` if you were talking about:
https://github.com/stylelint/stylelint-config-recommended/blob/93b5c713f9e1cf633f1c95ea88250526e2963117/index.js#L29

**Comment by clarfonthey** [comment]
> This sounds like a request for new a `depth` option. i.e. _type: new option_
Just to fully clarify: I mean not necessarily limiting the depth of the selector, just, only comparing selectors on the same depth. Namely:

```css
a {
  b {}
  c {}
}
d {
  e {}
  f {}
}
```

In this example, only the following pairs would be considered:
* `a` and `d`
* `b` and `c`
* `e` and `f`

> > 2. `no-descending-specificity` should expand the full selectors in its errors, and point out the line number for conflicting selectors.
> 
> That would be a _type: enhancement_.
Correct.

> > 3. `no-descending-specificity` should not be enabled by default.
> 
> Since I don't have access to your private registry I can't tell which config you are referring to. The issue would be transferred to `stylelint-config-recommended` if you were talking about:
You're right that I was referring to `stylelint-config-recommended`; I should have specified. In terms of the private registry, it's not a private registry, but regardless, the main rule in question is one that is included in stylelint by default.

I… wasn't aware that this was specifically documented as a limitation of the rule. I saw that it was working for my nested selectors, so, I assumed that it was designed to work with them, but I hadn't considered that it was just broken behaviour.

> 
> > And since I don't see any other issues pointing this out…
> 
> #4885 seems oddly similar.
My mistake; I only considered open issues, since this didn't seem to have been addressed at all. However, I hadn't considered that this was encouraged to be disabled when rules are nested.

----

I should add: I chose to use the new feature issue template for this because the existing templates required me to include a lot more information than I felt was necessary to describe the issue: that a lint provided by `stylelint` gives confusing output when enabled on nested rules. Since it's commonly used by other configs (like `stylelint-config-recommended`) it feels like it should have some sort of scrutiny applied to it.

That said, since `stylelint-config-recommended` isn't part of `stylelint`, I'll remove the note about it being enabled by default.

**Comment by romainmenke** [comment]
@clarfonthey Is this reproduction accurate:

https://stylelint.io/demo/#N4Igxg9gJgpiBcIACBbGUCWBDABAZzACcYYA7HYAHVMoBdSsA3C6uugMwkJRZtrdoAbLACMYg+OwyE8tALQR2c2gE8ADjF4CBAMngALCIxiF8arKXjws7Wia39HAyIK7wcxKAG5WTxwF9fAUCaPgFhMQlhWQUlVQ0HbVo8c0trW3sqML86FzccAHNiMh9s4KCAipC6EP8QABoQKUEYADksNAQQGAAPDrUWgDowPDwG8AhSKQKurJwcSm6eu1IoPEX3AG1qefnF2RUWwQxSeUgpjAK5WQsoLEIoRZ2cAF1qOsbz6YAxLhQsWhdABWeEm41gajGiDmCxAByOJ34CFhwjsskW9We+1UCNOci+l2utFu90eyMWqJg6JA7xA-iAA

-----

I don't personally think we should provide an option to limit by depth or to only warn about selectors at the same depth.

This would be very complex for us to provide and it doesn't add much value.

The feedback given by the rule is correct, it considers selectors in the same way browsers would do.

-----

> I… wasn't aware that this was specifically documented as a limitation of the rule. I saw that it was working for my nested selectors, so, I assumed that it was designed to work with them, but I hadn't considered that it was just broken behaviour.

A warning was indeed added for :

https://github.com/stylelint/stylelint/issues/5894


But if I try the same CSS again, I don't get any warnings with the latest version :

https://stylelint.io/demo/#N4Igxg9gJgpiBcIB0BnALgTwDYwLRgFd0IBbXFMAJwiywAJgAdAOzroDJ55cB3GAIwDWASzTkqNLPwCGlXGkrSwghizZsZygObUCzKPhoRK8OgHIMMWhB5mA3GroBfFo87c+Q0eOq0ZctAALAhJ+VVZ1TiU0YQA3GHD1dU1BHQg9A0gsY1MzShgoe0c2F2ZijnhAiHjKRKS6FLSMw2yTc34sAhgiiJLHUrd4aLiYABpBqpq6tndeARExCl8pWXlFZWnkpVTdfRac8x0YGGYepNK+5icQUZAAM2EcADlpEjhEGAAPV4AHHCQwCgUDdwBBmA8tAgQEwIowQF80CcoCg4aYANrlOHobBWYTMMSQcHCLTkNDSfSyKBwxwAXRY11uhIhADFjCRpGgoQArFBgkGwH7AxAwthYzA4LB4tCouhwrAcmDoOHjWEgbESqWGIkk9DkqCUmVyhVKkD0kBOIA

I think that maybe this rule used to have bugs, but is now fixed.

_Sidenote: Should we remove the warning?_

--------

> This is even a problem when it comes to nested selectors.

Yes, this is a problem :)

Nested selectors are just as prone to descending specificity as un-nested selectors.
The rule is giving accurate feedback here.

-------

> How you'd ever fix this issue.

I agree that this is unclear.
We don't currently have a good way to communicate two ranges that are somehow linked.

I think we should focus on trying to improve the warning message for cases like these.

**Comment by Mouvedia** [comment]
> > I… wasn't aware that this was specifically documented as a limitation of the rule. I saw that it was working for my nested selectors, so, I assumed that it was designed to work with them, but I hadn't considered that it was just broken behaviour.
> 
> A warning was indeed added for :
> 
> #5894
> 
> But if I try the same CSS again, I don't get any warnings with the latest version :
> 
> [stylelint.io/demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcIB0BnALgTwDYwLRgFd0IBbXFMAJwiywAJgAdAOzroDJ55cB3GAIwDWASzTkqNLPwCGlXGkrSwghizZsZygObUCzKPhoRK8OgHIMMWhB5mA3GroBfFo87c+Q0eOq0ZctAALAhJ+VVZ1TiU0YQA3GHD1dU1BHQg9A0gsY1MzShgoe0c2F2ZijnhAiHjKRKS6FLSMw2yTc34sAhgiiJLHUrd4aLiYABpBqpq6tndeARExCl8pWXlFZWnkpVTdfRac8x0YGGYepNK+5icQUZAAM2EcADlpEjhEGAAPV4AHHCQwCgUDdwBBmA8tAgQEwIowQF80CcoCg4aYANrlOHobBWYTMMSQcHCLTkNDSfSyKBwxwAXRY11uhIhADFjCRpGgoQArFBgkGwH7AxAwthYzA4LB4tCouhwrAcmDoOHjWEgbESqWGIkk9DkqCUmVyhVKkD0kBOIA)
> 
> I think that maybe this rule used to have bugs, but is now fixed.

#5897 only added a sentence in the README:

https://github.com/stylelint/stylelint/blob/1ba1a0aeafaa124a4987a0135d511111cec3af0a/lib/rules/no-descending-specificity/README.md?plain=1#L18

**Comment by romainmenke** [comment]
Filed a separate issue to see if we can improve the rule sufficiently to remove that warning: https://github.com/stylelint/stylelint/issues/7846

**Comment by clarfonthey** [comment]
> Would it be clearer with a message like this?
> 
> `Expected selector "span::after" to come before selector "nav { form { label:first-of-type { &:hover span::after } } }"`

Honestly, I personally would prefer just expanding both selectors so I can see the difference in specificity in the error itself, rather than still having to check the nesting for the rule manually.

**Comment by romainmenke** [comment]
In [Stylelint 17.0.0](https://stylelint.io/demo/#N4Igxg9gJgpiBcIACBbGUCWBDABAZzACcYYA7HYAHVMoBdSsA3C6uugMwkJRZtrdoAbLACMYg+OwyE8tALQR2c2gE8ADjF4CBAMngALCIxiF8arKXjws7Wia39HAyIK7wcxKAG5WTxwF9fAUCaPgFhMQlhWQUlVQ0HbVo8c0trW3sqML86FzccAHNiMh9s4KCAipC6EP8QABoQKUEYADksNAQQGAAPDrUWgDowPDwG8AhSKQKurJwcSm6eu1IoPEX3AG1qefnF2RUWwQxSeUgpjAK5WQsoLEIoRZ2cAF1qOsbz6YAxLhQsWhdQYHI4nWiEMCDABWeEm41gajGiDmCxAIPEYI2qOEdlki3qz32qlBpzkX0u11ot3ujwQ2IBMDxIHeIH8QA) the message has changed slightly

> 
Expected selector "span::after" to come before selector "nav form label:first-of-type:hover span::after", at line 5
(no-descending-specificity)

But I don't think we can consider this fully resolved yet.

------

In a way this isn't related to nesting at all:
https://stylelint.io/demo/#N4Igxg9gJgpiBcIB2BDAbgAgGYQE4FsMAbFAIxiPiwEtcBnAFwFoIsmGBPABxngAsIaGLgx0uKJPHgosDYRmAAdJIoaQieeBlwwoAbmUBfZctSYcBYmQrwSjFm049R4ydNnylKtRA24tOvpGIAA0IDREMAByKPhwiDAAHrFckQB0YHR0oeAQSDQA5gggXhgYiiBJckhQdBVaANrKZWUVjByRRNRIzJD51AVMjBJQKLhQFc0YALrBYX2FAGJ4+CgMxWntnd0MuGBpAFZ0eTmwXNmIpeUgWxQ79dckcowVIVNtnNs9TAsDQwwjMYTBCPNYwF4gYKGIA

```css
nav form label:first-of-type:hover span::after {
	color: red;
}

nav form label:last-of-type span::after {
	color: red;
}
```

> Expected selector "nav form label:last-of-type span::after" to come before selector "nav form label:first-of-type:hover span::after", at line 1
(no-descending-specificity)

Do not that `:first-of-type` and `:last-of-type` can match the same element when there is only one child.

------

Unsure how to proceed here so labeled it again as "needs discussion"
